presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java

```diff
@@ -97,6 +98,16 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
                     if (destinationColumnHandle == null) {
                         throw new PrestoException(COLUMN_NOT_FOUND, format("Did not find handle for column %s in destination table %s", destinationColumn, destinationTable));
                     }
+
+                    // validate that redirected types match source types
+                    Type sourceType = context.getSymbolAllocator().getTypes().get(entry.getKey());
+                    Type redirectedType = metadata.getColumnMetadata(context.getSession(), destinationTableHandle.get(), destinationColumnHandle).getType();
+                    if (!sourceType.equals(redirectedType)) {
+                        throw new PrestoException(
+                                TYPE_MISMATCH,
+                                format("Redirected type (%s) for column %s does not match source type (%s)", redirectedType, destinationColumn, sourceType));
```

Comment:
incluide source column handle as well?

Reply:
Please also mention source and target table. If a query uses multiple tables, knowing which one caused problems may be helpful.